conans/client/uploader.py

```diff
@@ -1,43 +1,92 @@
+    def __init__(self, paths, user_io, remote_proxy, search_manager, loader):
         self._paths = paths
         self._user_io = user_io
         self._remote_proxy = remote_proxy
+        self._search_manager = search_manager
+        self._loader = loader
 
-    def upload_conan(self, conan_ref, force=False, all_packages=False):
-        """Uploads the conans identified by conan_ref"""
-        export_path = self._paths.export(conan_ref)
-        if os.path.exists(export_path):
-            if not force:
-                self._check_package_date(conan_ref)
-
-            self._user_io.out.info("Uploading %s" % str(conan_ref))
-            self._remote_proxy.upload_conan(conan_ref)
-
-            if all_packages:
-                for index, package_id in enumerate(self._paths.conan_packages(conan_ref)):
-                    total = len(self._paths.conan_packages(conan_ref))
-                    self.upload_package(PackageReference(conan_ref, package_id), index + 1, total)
+    def upload_conan(self, pattern, force=False, all_packages=False, confirm=False,
+                     retry=None, retry_wait=None):
+        """Upload all the recipes matching 'pattern'"""
+        if is_a_reference(pattern):
+            ref = ConanFileReference.loads(pattern)
+            export_path = self._paths.export(ref)
+            if not os.path.exists(export_path):
+                raise NotFoundException("There is no local conanfile exported as %s"
+                                        % str(ref))
+            references = [ref, ]
+            confirm = True
         else:
-            self._user_io.out.error("There is no local conanfile exported as %s"
-                                    % str(conan_ref))
+            references = self._search_manager.search(pattern)
+
+        for conan_ref in references:
+            upload = True
+            if not confirm:
+                msg = "Are you sure you want to upload '%s'?" % str(conan_ref)
+                upload = self._user_io.request_boolean(msg)
+            if upload:
+                self._upload_conan(conan_ref, force, all_packages, retry, retry_wait)
+
+    def _upload_conan(self, conan_ref, force, all_packages, retry, retry_wait):
+        """Uploads the conans identified by conan_ref"""
+        if not force:
+            self._check_package_date(conan_ref)
 
-    def upload_package(self, package_ref, index=1, total=1):
+        self._user_io.out.info("Uploading %s" % str(conan_ref))
+        self._call_with_retry(retry, retry_wait, self._remote_proxy.upload_conan, conan_ref)
+
+        if all_packages:
+            for index, package_id in enumerate(self._paths.conan_packages(conan_ref)):
+                total = len(self._paths.conan_packages(conan_ref))
+                self.upload_package(PackageReference(conan_ref, package_id), index + 1, total,
+                                    retry, retry_wait)
+
+    def _check_reference(self, conan_reference):
```

Comment:
I don't like that this is repeated for every binary, even if it is common code, only for the recipe.